delivery_send_to_shipper_at_operation/tests/test_send_to_shipper.py

```diff
@@ -0,0 +1,129 @@
+                "product_id": delivery_fee.id,
+                "integration_level": "rate_and_ship",
+                "send_delivery_notice_on": "ship",
+            }
+        )
+        cls.carrier_on_pack = cls.env["delivery.carrier"].create(
+            {
+                "name": "Carrier On Pack",
+                "product_id": delivery_fee.id,
+                "integration_level": "rate_and_ship",
+                "send_delivery_notice_on": "custom",
+                "send_delivery_notice_picking_type_ids": [
+                    (6, 0, cls.wh.pack_type_id.ids)
+                ],
+            }
+        )
+        # create a pick/pack/ship transfer
+        cls.ship_move = cls.env["stock.move"].create(
+            {
+                "name": cls.product.display_name,
+                "product_id": cls.product.id,
+                "product_uom_qty": 10.0,
+                "product_uom": cls.product.uom_id.id,
+                "location_id": cls.ship_location.id,
+                "location_dest_id": cls.customer_location.id,
+                "warehouse_id": cls.wh.id,
+                "picking_type_id": cls.wh.out_type_id.id,
+                "procure_method": "make_to_order",
+                "state": "draft",
+            }
+        )
+        cls.ship_move._assign_picking()
+        cls.ship_move._action_confirm()
+        cls.pack_move = cls.ship_move.move_orig_ids[0]
+        cls.pick_move = cls.pack_move.move_orig_ids[0]
+        cls.picking = cls.pick_move.picking_id
+        cls.packing = cls.pack_move.picking_id
+        cls.shipping = cls.ship_move.picking_id
+        cls.picking.action_assign()
+
+    def _validate_picking(self, picking):
+        for ml in picking.move_line_ids:
+            ml.qty_done = ml.product_uom_qty
+        picking.action_done()
+
+    def test_send_to_shipper_on_ship(self):
+        """Check sending of delivery notification on ship.
+
+        No delivery notification is sent to the carrier when the pack is
+        validated (std Odoo behavior).
+        """
+        self.shipping.carrier_id = self.carrier_on_ship
+        self._validate_picking(self.picking)
+        self.assertEqual(self.picking.state, "done")
+        # Validate the pack: delivery notification is not sent
+        with mock.patch.object(type(self.packing), "send_to_shipper") as mocked:
+            self._validate_picking(self.packing)
+            self.assertEqual(self.packing.state, "done")
+            mocked.assert_not_called()
+            self.assertFalse(self.shipping.delivery_notification_sent)
+        # Validate the ship: delivery notification is sent
+        with mock.patch.object(
+            type(self.shipping.carrier_id), "send_shipping"
+        ) as mocked:
+            self._validate_picking(self.shipping)
+            self.assertEqual(self.shipping.state, "done")
+            mocked.assert_called()
+            self.assertFalse(self.shipping.delivery_notification_sent)
```

Comment:
I would expect the flag to be true here. Or, do we expect to be able to send it multiple times on the ship?

Reply:
I was wondering the same, due to a lack of time I didn't do it: the thing is, right now with the default config we do not change at all the std behavior, but yes we could set this field in all cases. This would allows to filter operations from which we sent a delivery notification.

Reply:
We can not do that easily: if we set the `delivery_notification_sent` on a ship (with default Odoo behavior), the check done in `send_to_shipper` on this field will remove the possibility to call again this method to re-send the notification to the carrier, we should avoid that IMO (even if the button is removed from the form as soon as a tracking ref has been retrieved from the carrier, technically we should still be able to call this method).